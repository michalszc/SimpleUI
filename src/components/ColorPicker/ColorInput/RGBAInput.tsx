import React, { FC, useCallback, useMemo } from "react";
import * as CSS from "csstype";
import CustomInput from "./CustomInput";
import styled from "styled-components";
import { hexToRgba, rgbaToHex } from "../../../utils";
import { floor, isNil, mapValues } from "lodash";
import Styles from "../../../utils/styles";

export interface RGBAInputProps {
    color: CSS.DataType.Color;
    setColor: (value: CSS.DataType.Color) => void;
}

interface RGBA {
    r?: number;
    g?: number;
    b?: number;
    a?: number;
}

const RGBAContainer = styled.div`
    display: grid;
    grid-template-columns: repeat(4, minmax(25%, 1fr));
    width: inherit;
    margin: 0 auto;
    & > .simpleui-colorpicker-inputcontainer {
        padding: ${ props => props.theme.sizes[2] };
    }
    & > .simpleui-colorpicker-inputcontainer:not(:first-of-type) {
        padding-left: ${ props => props.theme.sizes[.5] };
    }
    & > .simpleui-colorpicker-inputcontainer:not(:last-of-type) {
        padding-right: ${ props => props.theme.sizes[.5] };
    }
`;

const RGBAInput: FC<RGBAInputProps> = ({
    color, setColor
}) => {
    const rgbaColor = useMemo(
        () => {
            const _rgba = hexToRgba(color);
            if (isNil(_rgba)) {
                return {
                    r: '0',
                    g: '0',
                    b: '0',
                    a: '0'
                };
            } else {
                return mapValues(_rgba, (v: number) => v.toString());
            }
        },
        [color]
    );

    const setNewColor = useCallback(
        (rgba: RGBA) => {
            const _rgba = {
                ...mapValues(rgbaColor, (v: string) => Number(v)),
                ...mapValues(rgba, (v: number, k: string) => {
                    if (k !== 'a') {
                        return (v > 255) ? floor(v / 10) : v;
                    } else {
                        return (v > 100) ? floor(v / 10) : v;
                    }
                }) as RGBA
            };
            setColor(
                rgbaToHex(
                    _rgba.r,
                    _rgba.g,
                    _rgba.b,
                    _rgba.a
                )
            );
        }, [rgbaColor]);

    return (
        <Styles>
            <RGBAContainer className={"simpleui-colorpicker-rgbacontainer"}>
                <CustomInput prefix={'R'} value={rgbaColor.r} onChange={(v: string) => setNewColor({
                    r: Number(v)
                })} />
                <CustomInput prefix={'G'} value={rgbaColor.g} onChange={(v: string) => setNewColor({
                    g: Number(v)
                })} />
                <CustomInput prefix={'B'} value={rgbaColor.b} onChange={(v: string) => setNewColor({
                    b: Number(v)
                })} />
                <CustomInput prefix={'A'} value={rgbaColor.a} onChange={(v: string) => setNewColor({
                    a: Number(v)
                })} />
            </RGBAContainer>
        </Styles>
    );
};

export default RGBAInput;
