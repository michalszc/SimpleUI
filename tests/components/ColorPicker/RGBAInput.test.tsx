import React from "react";
import '@testing-library/jest-dom';
import { render } from "@testing-library/react";
import RGBAInput from "../../../src/components/ColorPicker/ColorInput/RGBAInput";

describe("RGBAInput", () => {
    const color = '#FFF';
    const mockFn = jest.fn();
    test("should properly render RGBAInput component", () => {
        const container = render(<RGBAInput color={color} setColor={mockFn}/>);
        expect(container).toMatchSnapshot();
    });
    test("should properly set color values", () => {
        const { getAllByRole } = render(<RGBAInput color={color} setColor={mockFn}/>);
        const inputs = getAllByRole('textbox');
        expect(inputs[0]).toHaveValue('255');
        expect(inputs[1]).toHaveValue('255');
        expect(inputs[2]).toHaveValue('255');
        expect(inputs[3]).toHaveValue('100');
    });
});
