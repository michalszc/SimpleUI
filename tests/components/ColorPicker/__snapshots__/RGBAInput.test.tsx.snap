// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RGBAInput should properly render RGBAInput component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-hLBbgP epUoVT simpleui-colorpicker-rgbacontainer"
        mode="light"
      >
        <div
          class="sc-bcXHqe dLNOCe simpleui-colorpicker-inputcontainer"
          mode="light"
        >
          <div
            class="sc-gswNZR icpVtD simpleui-colorpicker-input-prefix"
          >
            R
          </div>
          <input
            class="sc-dkrFOg dGsRDm simpleui-colorpicker-input"
            value="255"
          />
        </div>
        <div
          class="sc-bcXHqe dLNOCe simpleui-colorpicker-inputcontainer"
          mode="light"
        >
          <div
            class="sc-gswNZR icpVtD simpleui-colorpicker-input-prefix"
          >
            G
          </div>
          <input
            class="sc-dkrFOg dGsRDm simpleui-colorpicker-input"
            value="255"
          />
        </div>
        <div
          class="sc-bcXHqe dLNOCe simpleui-colorpicker-inputcontainer"
          mode="light"
        >
          <div
            class="sc-gswNZR icpVtD simpleui-colorpicker-input-prefix"
          >
            B
          </div>
          <input
            class="sc-dkrFOg dGsRDm simpleui-colorpicker-input"
            value="255"
          />
        </div>
        <div
          class="sc-bcXHqe dLNOCe simpleui-colorpicker-inputcontainer"
          mode="light"
        >
          <div
            class="sc-gswNZR icpVtD simpleui-colorpicker-input-prefix"
          >
            A
          </div>
          <input
            class="sc-dkrFOg dGsRDm simpleui-colorpicker-input"
            value="100"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-hLBbgP epUoVT simpleui-colorpicker-rgbacontainer"
      mode="light"
    >
      <div
        class="sc-bcXHqe dLNOCe simpleui-colorpicker-inputcontainer"
        mode="light"
      >
        <div
          class="sc-gswNZR icpVtD simpleui-colorpicker-input-prefix"
        >
          R
        </div>
        <input
          class="sc-dkrFOg dGsRDm simpleui-colorpicker-input"
          value="255"
        />
      </div>
      <div
        class="sc-bcXHqe dLNOCe simpleui-colorpicker-inputcontainer"
        mode="light"
      >
        <div
          class="sc-gswNZR icpVtD simpleui-colorpicker-input-prefix"
        >
          G
        </div>
        <input
          class="sc-dkrFOg dGsRDm simpleui-colorpicker-input"
          value="255"
        />
      </div>
      <div
        class="sc-bcXHqe dLNOCe simpleui-colorpicker-inputcontainer"
        mode="light"
      >
        <div
          class="sc-gswNZR icpVtD simpleui-colorpicker-input-prefix"
        >
          B
        </div>
        <input
          class="sc-dkrFOg dGsRDm simpleui-colorpicker-input"
          value="255"
        />
      </div>
      <div
        class="sc-bcXHqe dLNOCe simpleui-colorpicker-inputcontainer"
        mode="light"
      >
        <div
          class="sc-gswNZR icpVtD simpleui-colorpicker-input-prefix"
        >
          A
        </div>
        <input
          class="sc-dkrFOg dGsRDm simpleui-colorpicker-input"
          value="100"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
